services/code_hosting/git_hub_service.rb

```diff
@@ -1,118 +1,352 @@
+      client.pull_commits(self.project.repo_config.full_name, number).map(&:sha)
+    end
+
+    # Returns the last sha for every pull request that targets a list of branches (or all if not given) and status (or open if not given)
+    # @param [Array<String>, nil] branches, Either an array of target branches names or nil.
+    # @param [String] state, Either open, closed, or all to filter by state. Default: open.
+    # @return [Array<String>] Array of the last commit SHA for the given repo and state.
+    def last_commit_sha_for_pull_requests(branches: nil, state: "open")
+      pull_requests_urls = self.pull_requests(branches: branches, state: state)
+      numbers = pull_requests_urls.map { |url| URI.parse(url) }.map(&:path).collect { |path| path.split("/").last }
+      return numbers.map { |number| commits_sha_from_pull_request(number: number) }.map(&:last)
     end
 
     # returns the statused of a given commit sha for a given repo specifically for fastlane.ci
     # TODO: add support for filtering status types, to allow listing of just fastlane.ci status reports
     #       This has to wait for now, until we decide how we separate them for each project, as multiple projects
     #       can run builds for one repo
-    def statuses_for_commit_sha(repo_full_name: nil, sha: nil)
-      all_statuses = client.statuses(repo_full_name, sha)
-      only_ci_statuses = all_statuses.select { |status| status.context.start_with?(GitHubService.status_context_prefix) }
+    def statuses_for_commit_sha(sha: nil)
+      all_statuses = client.statuses(self.project.repo_config.full_name, sha)
+      only_ci_statuses = all_statuses.select { |status| status.context == self.status_context }
       return only_ci_statuses
     end
 
     # updates the most current commit to "pending" on all open prs if they don't have a status.
     # returns a list of commits that have been updated to `pending` status
-    def update_all_open_prs_without_status_to_pending_status!(repo_full_name: nil, status_context: nil)
-      open_pr_commits = self.last_commit_sha_for_all_open_pull_requests(repo_full_name: repo_full_name)
+    def update_all_open_prs_without_status_to_pending_status!
+      open_pr_commits = self.last_commit_sha_for_all_open_pull_requests
       updated_commits = []
       open_pr_commits.each do |sha|
-        statuses = self.statuses_for_commit_sha(repo_full_name: repo_full_name, sha: sha)
+        statuses = self.statuses_for_commit_sha(sha: sha)
         if statuses.count == 0
-          self.set_build_status!(repo: repo_full_name, sha: sha, state: "pending", status_context: status_context)
+          self.set_build_status!(sha: sha, state: "pending")
           updated_commits << sha
         end
       end
       return updated_commits
     end
 
-    # TODO: parse those here or in service layer?
-    def repos
-      client.repos({}, query: { sort: "asc" })
+    # @return [Array<String>]
+    def branches
```

Comment:
rename `branch_names`

Reply:
I kinda like `branches` for simplicity, as there's no confusion with what other thing might be, do you prefer `branch_names` tho? I mean, a branch itself it's (basically) his name, so I thought it was a little verbose, but can refactor if you still want to 😄

Reply:
The reason why I like `branch_names` is because `branch` is an object with metadata, and you're mapping an array of `branch` objects to an array of `name` strings. So I think for clarity, to show that these are no longer branch objects but instead, they are the names of each branch, renaming to `branch_names` is a good idea. 

I often times think things can seem verbose, but if you've never seen the code, that verbosity can help with understanding. So I actually prefer verbosity over conciseness. It usually lends itself to more maintainable code 😄